0003-for-each.md

```diff
@@ -0,0 +1,634 @@
+- Feature Name: for_each_loops
+- Start Date: 2022-02-22 (twosday!)
+- RFC PR: [dafny-lang/rfcs#9](https://github.com/dafny-lang/rfcs/pull/9)
+- Dafny Issue: [dafny-lang/dafny#1753](https://github.com/dafny-lang/dafny/issues/1753)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes adding two closely-related new features to Dafny for working with collections :
+
+1. `foreach` loops over any enumerable domain, initially only supporting the builtin collection types but leaving the door open for user-defined collections.
+2. Generalized sequence comprehension expressions, similar to the existing `set` and `map` comprehension expressions and more flexible than the current `seq(5, i => i*i)` form.
+
+# Motivation
+[motivation]: #motivation
+
+Loops are notoriously difficult for Dafny users to verify, and coming up with the correct loop invariants to prove a loop correct 
+is particularly challenging. The recent addition of `for` loops was a good step to abstract away from raw `while` loops,
+but only supports consecutive integer indices. They still fall short of the expressiveness and economy of `for` or `foreach` loops in
+many mainstream programming languages, which generally iterate through the contents of a collection of some kind.
+It is a well-established best practice to avoid manual loop indexes wherever possible, as they force the assumption that datatypes
+support efficient random access, and are more likely to be used incorrectly by accident:
+
+```dafny
+// Before:
+method AllPairs(s: seq<nat>) returns (result: seq<(nat, nat)>) {
+  for i := 0 to |s| {
+    var left := s[i];
+    for j := 0 to |s| {
+      var right := s[i]; // Whoops!
+      result := result + (left, right);  
+    }
+  }
+}
+
+// After:
+method AllPairs(s: seq<nat>) returns (result: seq<(nat, nat)>) {
+  foreach left | left in s {
+    foreach right | right in s {
+      result := result + (left, right);
+    }
+  }
+}
+
+// Or even better, as an expression:
+function AllPairs(s: seq<nat>): seq<(nat, nat)> {
+  seq left, right | left in s && right in s
+}
```

Comment:
Good point, I'll put some thought into improving the motivating example - I think even the eventual user-facing documentation should have a good example of where a `foreach` loop is necessary because a sequence comprehension doesn't cut it.

Reply:
Yeah, it seems to me that, even if `foreach` and sequence comprehensions were identical in what they could express, `foreach` would feel a lot more natural if the body is very imperative.

Reply:
This is part of why it feels right to me to design both features at once, as they are kind of two sides of the same fundamental coin: imperative and functional.